File: src/NullDemo.java
import java.util.HashMap;
import java.util.Map;

public class NullDemo {
    static String str = null;

    public static void main(String[] args) {
        System.out.println(str.length());
        System.out.println(getTranslations().get("four"));

        // subtle, but could fail due to async
        if (str != null) {
            int x = str.length();
        }
    }

    private static Map<String, String> getTranslations() {
        HashMap<String, String> translations = new HashMap<>();
        translations.put("one", "ichi");
        translations.put("two", "ni");
        translations.put("three", "san");
        return translations;
    }
}

File: src/NullDemo.kt
// :o doesn't compile val str: String = null
var str: String? = null
val strImmutable: String? = null

fun main() {
    // println(str.length)
    println(str?.length)
    println(str?.length ?: 0)
    // will crash println(str!!.length)
    println(getTranslations()["four"])

    if (strImmutable != null) {
        val x = strImmutable.length * 2
    }

//    doesn't compile
//    if (str != null) {
//        val x = str.length * 2
//    }

    str?.let { str ->
        val x = str.length * 2
    }

    // can also turn stuff into null conditionally!


}

fun getTranslations() = mapOf("one" to "ichi", "two" to "ni", "three" to "san")